File: tetra/bash/midi/tmc.js
#!/usr/bin/env node

/**
 * TMC Node.js Bridge - Bidirectional MIDI I/O using easymidi
 * Alternative to tmc.c for easier cross-platform installation
 */

const net = require('net');
const easymidi = require('easymidi');

class TMCBridge {
    constructor(options = {}) {
        this.socketPath = options.socketPath;
        this.inputDeviceId = options.inputDevice ?? -1;
        this.outputDeviceId = options.outputDevice ?? -1;
        this.verbose = options.verbose || false;
        this.running = true;

        this.midiInput = null;
        this.midiOutput = null;
        this.socket = null;
    }

    log(msg) {
        if (this.verbose) {
            console.error(`[TMC] ${msg}`);
        }
    }

    listDevices() {
        console.log('Available MIDI Devices:');
        console.log('========================\n');

        console.log('Input Devices:');
        const inputs = easymidi.getInputs();
        inputs.forEach((name, i) => {
            console.log(`  [${i}] ${name}`);
        });

        console.log('\nOutput Devices:');
        const outputs = easymidi.getOutputs();
        outputs.forEach((name, i) => {
            console.log(`  [${i}] ${name}`);
        });
        console.log();
    }

    initMIDI() {
        const inputs = easymidi.getInputs();
        const outputs = easymidi.getOutputs();

        // Initialize input
        if (this.inputDeviceId < 0 && inputs.length > 0) {
            this.inputDeviceId = 0;
        }

        if (this.inputDeviceId >= 0 && this.inputDeviceId < inputs.length) {
            const inputName = inputs[this.inputDeviceId];
            this.midiInput = new easymidi.Input(inputName);
            this.log(`Opened MIDI input: ${inputName}`);

            // Register event handlers
            this.midiInput.on('noteon', (msg) => this.handleMidiEvent('noteon', msg));
            this.midiInput.on('noteoff', (msg) => this.handleMidiEvent('noteoff', msg));
            this.midiInput.on('cc', (msg) => this.handleMidiEvent('cc', msg));
            this.midiInput.on('program', (msg) => this.handleMidiEvent('program', msg));
            this.midiInput.on('pitchbend', (msg) => this.handleMidiEvent('pitchbend', msg));
        } else if (this.inputDeviceId >= 0) {
            console.error(`ERROR: Input device ${this.inputDeviceId} not found`);
        }

        // Initialize output
        if (this.outputDeviceId < 0 && outputs.length > 0) {
            this.outputDeviceId = 0;
        }

        if (this.outputDeviceId >= 0 && this.outputDeviceId < outputs.length) {
            const outputName = outputs[this.outputDeviceId];
            this.midiOutput = new easymidi.Output(outputName);
            this.log(`Opened MIDI output: ${outputName}`);
        } else if (this.outputDeviceId >= 0) {
            console.error(`ERROR: Output device ${this.outputDeviceId} not found`);
        }
    }

    handleMidiEvent(type, msg) {
        let formatted = null;

        switch (type) {
            case 'noteon':
                if (msg.velocity === 0) {
                    formatted = `NOTE_OFF ${msg.channel + 1} ${msg.note}\n`;
                } else {
                    formatted = `NOTE_ON ${msg.channel + 1} ${msg.note} ${msg.velocity}\n`;
                }
                break;

            case 'noteoff':
                formatted = `NOTE_OFF ${msg.channel + 1} ${msg.note}\n`;
                break;

            case 'cc':
                formatted = `CC ${msg.channel + 1} ${msg.controller} ${msg.value}\n`;
                break;

            case 'program':
                formatted = `PROGRAM_CHANGE ${msg.channel + 1} ${msg.number}\n`;
                break;

            case 'pitchbend':
                formatted = `PITCH_BEND ${msg.channel + 1} ${msg.value}\n`;
                break;
        }

        if (formatted) {
            if (this.verbose) {
                process.stderr.write(`MIDI IN: ${formatted}`);
            }

            if (this.socket) {
                this.socket.write(formatted);
            }
        }
    }

    parseMidiCommand(line) {
        const parts = line.trim().split(/\s+/);
        if (parts.length === 0 || !this.midiOutput) {
            return;
        }

        const cmd = parts[0];

        try {
            switch (cmd) {
                case 'CC':
                    if (parts.length >= 4) {
                        this.midiOutput.send('cc', {
                            channel: parseInt(parts[1]) - 1,
                            controller: parseInt(parts[2]),
                            value: parseInt(parts[3])
                        });
                        if (this.verbose) {
                            process.stderr.write(`MIDI OUT: ${line}\n`);
                        }
                    }
                    break;

                case 'NOTE_ON':
                    if (parts.length >= 4) {
                        this.midiOutput.send('noteon', {
                            channel: parseInt(parts[1]) - 1,
                            note: parseInt(parts[2]),
                            velocity: parseInt(parts[3])
                        });
                        if (this.verbose) {
                            process.stderr.write(`MIDI OUT: ${line}\n`);
                        }
                    }
                    break;

                case 'NOTE_OFF':
                    if (parts.length >= 3) {
                        this.midiOutput.send('noteoff', {
                            channel: parseInt(parts[1]) - 1,
                            note: parseInt(parts[2]),
                            velocity: 0
                        });
                        if (this.verbose) {
                            process.stderr.write(`MIDI OUT: ${line}\n`);
                        }
                    }
                    break;

                case 'PROGRAM_CHANGE':
                    if (parts.length >= 3) {
                        this.midiOutput.send('program', {
                            channel: parseInt(parts[1]) - 1,
                            number: parseInt(parts[2])
                        });
                        if (this.verbose) {
                            process.stderr.write(`MIDI OUT: ${line}\n`);
                        }
                    }
                    break;
            }
        } catch (err) {
            console.error(`ERROR: Failed to send MIDI: ${err.message}`);
        }
    }

    connectSocket() {
        return new Promise((resolve, reject) => {
            this.socket = net.createConnection(this.socketPath, () => {
                this.log(`Connected to socket: ${this.socketPath}`);
                resolve();
            });

            this.socket.on('error', (err) => {
                console.error(`ERROR: Socket error: ${err.message}`);
                reject(err);
            });

            this.socket.on('close', () => {
                this.log('Socket closed');
                this.running = false;
            });

            let buffer = '';
            this.socket.on('data', (data) => {
                buffer += data.toString();
                const lines = buffer.split('\n');
                buffer = lines.pop(); // Keep incomplete line in buffer

                lines.forEach(line => {
                    if (line.trim()) {
                        this.parseMidiCommand(line);
                    }
                });
            });
        });
    }

    async run() {
        this.log('Starting TMC Bridge...');

        this.initMIDI();

        try {
            await this.connectSocket();
        } catch (err) {
            console.error(`ERROR: Failed to connect to socket: ${err.message}`);
            process.exit(1);
        }

        console.log('TMC Bridge running (Ctrl+C to stop)');

        // Handle graceful shutdown
        process.on('SIGINT', () => {
            console.log('\nShutting down...');
            this.cleanup();
            process.exit(0);
        });

        process.on('SIGTERM', () => {
            this.cleanup();
            process.exit(0);
        });
    }

    cleanup() {
        this.running = false;

        if (this.midiInput) {
            this.midiInput.close();
        }
        if (this.midiOutput) {
            this.midiOutput.close();
        }
        if (this.socket) {
            this.socket.end();
        }
    }
}

// CLI
function main() {
    const args = process.argv.slice(2);
    const options = {
        inputDevice: -1,
        outputDevice: -1,
        socketPath: null,
        verbose: false
    };

    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '-l':
            case '--list':
                const bridge = new TMCBridge();
                bridge.listDevices();
                process.exit(0);
                break;

            case '-i':
            case '--input':
                options.inputDevice = parseInt(args[++i]);
                break;

            case '-o':
            case '--output':
                options.outputDevice = parseInt(args[++i]);
                break;

            case '-s':
            case '--socket':
                options.socketPath = args[++i];
                break;

            case '-v':
            case '--verbose':
                options.verbose = true;
                break;

            case '-h':
            case '--help':
                console.log(`
Usage: tmc.js [OPTIONS]

Options:
  -l, --list              List available MIDI devices and exit
  -i, --input DEVICE_ID   MIDI input device ID
  -o, --output DEVICE_ID  MIDI output device ID
  -s, --socket PATH       Unix socket path for TMC communication
  -v, --verbose           Verbose output
  -h, --help              Show this help

Example:
  node tmc.js -i 0 -o 0 -s /tmp/tmc.sock -v
`);
                process.exit(0);
                break;
        }
    }

    if (!options.socketPath) {
        console.error('ERROR: Socket path required (-s option)');
        process.exit(1);
    }

    const bridge = new TMCBridge(options);
    bridge.run();
}

main();

File: tetra/bash/midi/tmc.py
#!/usr/bin/env python3
"""
TMC Python Bridge - Bidirectional MIDI I/O using python-rtmidi
Alternative to tmc.c for easier installation on Apple Silicon
"""

import sys
import socket
import threading
import time
import argparse
from pathlib import Path

try:
    import rtmidi
except ImportError:
    print("Error: python-rtmidi not installed")
    print("Install with: pip3 install python-rtmidi")
    sys.exit(1)


class TMCBridge:
    def __init__(self, socket_path, input_device=-1, output_device=-1, verbose=False):
        self.socket_path = socket_path
        self.input_device_id = input_device
        self.output_device_id = output_device
        self.verbose = verbose
        self.running = True

        self.midi_in = None
        self.midi_out = None
        self.sock = None

    def log(self, msg):
        if self.verbose:
            print(f"[TMC] {msg}", file=sys.stderr)

    def list_devices(self):
        """List available MIDI devices"""
        print("Available MIDI Devices:")
        print("========================\n")

        print("Input Devices:")
        midi_in = rtmidi.MidiIn()
        for i, port in enumerate(midi_in.get_ports()):
            print(f"  [{i}] {port}")

        print("\nOutput Devices:")
        midi_out = rtmidi.MidiOut()
        for i, port in enumerate(midi_out.get_ports()):
            print(f"  [{i}] {port}")
        print()

    def init_midi(self):
        """Initialize MIDI devices"""
        # Initialize input
        self.midi_in = rtmidi.MidiIn()
        available_inputs = self.midi_in.get_ports()

        if self.input_device_id < 0:
            # Auto-detect first available
            if available_inputs:
                self.input_device_id = 0
            else:
                self.log("No MIDI input devices found")
                self.midi_in = None

        if self.midi_in and self.input_device_id < len(available_inputs):
            self.midi_in.open_port(self.input_device_id)
            self.midi_in.set_callback(self.midi_callback)
            self.log(f"Opened MIDI input: {available_inputs[self.input_device_id]}")

        # Initialize output
        self.midi_out = rtmidi.MidiOut()
        available_outputs = self.midi_out.get_ports()

        if self.output_device_id < 0:
            # Auto-detect first available
            if available_outputs:
                self.output_device_id = 0
            else:
                self.log("No MIDI output devices found")
                self.midi_out = None

        if self.midi_out and self.output_device_id < len(available_outputs):
            self.midi_out.open_port(self.output_device_id)
            self.log(f"Opened MIDI output: {available_outputs[self.output_device_id]}")

    def connect_socket(self):
        """Connect to Unix socket"""
        self.sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        try:
            self.sock.connect(self.socket_path)
            self.log(f"Connected to socket: {self.socket_path}")
        except Exception as e:
            print(f"ERROR: Failed to connect to socket: {e}", file=sys.stderr)
            sys.exit(1)

    def format_midi_message(self, message):
        """Format MIDI message for socket"""
        if not message:
            return None

        status = message[0]
        msg_type = status & 0xF0
        channel = (status & 0x0F) + 1  # 1-indexed

        if msg_type == 0x80:  # Note Off
            return f"NOTE_OFF {channel} {message[1]}\n"
        elif msg_type == 0x90:  # Note On
            if len(message) > 2 and message[2] == 0:
                return f"NOTE_OFF {channel} {message[1]}\n"
            else:
                return f"NOTE_ON {channel} {message[1]} {message[2]}\n"
        elif msg_type == 0xB0:  # Control Change
            return f"CC {channel} {message[1]} {message[2]}\n"
        elif msg_type == 0xC0:  # Program Change
            return f"PROGRAM_CHANGE {channel} {message[1]}\n"
        elif msg_type == 0xE0:  # Pitch Bend
            bend = (message[2] << 7) | message[1]
            return f"PITCH_BEND {channel} {bend}\n"
        else:
            return f"UNKNOWN {status:02X} {message[1] if len(message) > 1 else 0} {message[2] if len(message) > 2 else 0}\n"

    def midi_callback(self, event, data=None):
        """Callback for incoming MIDI messages"""
        message, deltatime = event
        formatted = self.format_midi_message(message)

        if formatted:
            if self.verbose:
                print(f"MIDI IN: {formatted.strip()}")

            if self.sock:
                try:
                    self.sock.sendall(formatted.encode())
                except Exception as e:
                    print(f"ERROR: Failed to send to socket: {e}", file=sys.stderr)
                    self.running = False

    def parse_midi_command(self, line):
        """Parse socket command and send MIDI output"""
        parts = line.strip().split()
        if not parts or not self.midi_out:
            return

        cmd = parts[0]

        try:
            if cmd == "CC" and len(parts) >= 4:
                channel = int(parts[1]) - 1
                controller = int(parts[2])
                value = int(parts[3])
                self.midi_out.send_message([0xB0 | (channel & 0x0F), controller & 0x7F, value & 0x7F])
                if self.verbose:
                    print(f"MIDI OUT: {line.strip()}")

            elif cmd == "NOTE_ON" and len(parts) >= 4:
                channel = int(parts[1]) - 1
                note = int(parts[2])
                velocity = int(parts[3])
                self.midi_out.send_message([0x90 | (channel & 0x0F), note & 0x7F, velocity & 0x7F])
                if self.verbose:
                    print(f"MIDI OUT: {line.strip()}")

            elif cmd == "NOTE_OFF" and len(parts) >= 3:
                channel = int(parts[1]) - 1
                note = int(parts[2])
                self.midi_out.send_message([0x80 | (channel & 0x0F), note & 0x7F, 0])
                if self.verbose:
                    print(f"MIDI OUT: {line.strip()}")

            elif cmd == "PROGRAM_CHANGE" and len(parts) >= 3:
                channel = int(parts[1]) - 1
                program = int(parts[2])
                self.midi_out.send_message([0xC0 | (channel & 0x0F), program & 0x7F])
                if self.verbose:
                    print(f"MIDI OUT: {line.strip()}")
        except Exception as e:
            print(f"ERROR: Failed to send MIDI: {e}", file=sys.stderr)

    def socket_listener(self):
        """Listen for commands from socket"""
        buffer = ""
        while self.running:
            try:
                data = self.sock.recv(1024)
                if not data:
                    break

                buffer += data.decode()
                while '\n' in buffer:
                    line, buffer = buffer.split('\n', 1)
                    if line:
                        self.parse_midi_command(line)
            except Exception as e:
                print(f"ERROR: Socket read error: {e}", file=sys.stderr)
                break

    def run(self):
        """Main run loop"""
        self.log("Starting TMC Bridge...")

        self.init_midi()
        self.connect_socket()

        if self.midi_out:
            # Start socket listener thread
            listener_thread = threading.Thread(target=self.socket_listener, daemon=True)
            listener_thread.start()

        print("TMC Bridge running (Ctrl+C to stop)")

        try:
            while self.running:
                time.sleep(1)
        except KeyboardInterrupt:
            print("\nShutting down...")
        finally:
            self.cleanup()

    def cleanup(self):
        """Clean up resources"""
        self.running = False

        if self.midi_in:
            self.midi_in.close_port()
        if self.midi_out:
            self.midi_out.close_port()
        if self.sock:
            self.sock.close()


def main():
    parser = argparse.ArgumentParser(description="TMC Bridge - MIDI I/O for Tetra")
    parser.add_argument('-l', '--list', action='store_true',
                       help='List available MIDI devices and exit')
    parser.add_argument('-i', '--input', type=int, default=-1,
                       help='MIDI input device ID')
    parser.add_argument('-o', '--output', type=int, default=-1,
                       help='MIDI output device ID')
    parser.add_argument('-s', '--socket', type=str, required=False,
                       help='Unix socket path for TMC communication')
    parser.add_argument('-v', '--verbose', action='store_true',
                       help='Verbose output')

    args = parser.parse_args()

    if args.list:
        bridge = TMCBridge("", verbose=args.verbose)
        bridge.list_devices()
        return 0

    if not args.socket:
        print("ERROR: Socket path required (-s option)", file=sys.stderr)
        parser.print_help()
        return 1

    bridge = TMCBridge(args.socket, args.input, args.output, args.verbose)
    bridge.run()

    return 0


if __name__ == "__main__":
    sys.exit(main())
